### DISCUSSION TITLE
Raspberry USB speed setup ?

### DISCUSSION THREAD

**Discussion by todely** [awawa-dev/HyperSerialEsp8266#7] [Q&A] [answered]
Hello,

I had a setup (raspberry + arduino) with an old led strip and change led strip for a SK6812 RGBW.
Problem wih fastLED on arduino (no rgbw available..)

So i think i can replace my arduino with a wemos D1 mini, connected to my raspberry and driving led strip with PIN D4 right ?

Question:
I had setup USB speed of raspberry to 500000 to exchange with arduino and I see I need to setup @2000000 but can't find / remember where on my raspberry (not in hyperion setup, on Raspberry) i can change USB speed ?

Thanks

**Comment by awawa-dev** [comment, marked as answer]
OK, some things first
1) your Arduino has 5v GPIO which is just fine for sk6812. If Arduino wouldn't be so slow it could be a good solution. After you switch to Esp8266 (or Esp32 or connect directly to Rpi's GPIO) then you have 3.3V GPIO. Just a warning: may work or you may need a voltage level shifter (SN74AHCT125N).
2) if you want to use 2Mb speed you must make sure that you have 2Mb capable serial chip like ch340g (big rectangle chip with a label, there is a lot of counterfeits) or cp2104 on your ESP board. Inspect your wemos D1 mini. The only thing that you need to set up in HyperHDR is the adalight driver (BTW these settings don't exist in Hyperion 😉 ):
![](https://user-images.githubusercontent.com/69086569/130987180-f12fb88e-850a-4fc3-8ea8-3579d900059a.png)

**Comment by todely** [comment]
Thanks for your answer !

1/ Could i stay with my Arduino ? (but i need the right sketch and can't find it...)
2/ Baudrate settings exist in Hyperion, i have put 500000 and same in my arduino fastled sktech but where is the parameter for raspbian to change limit ?
**EDIT: seems to be in config.txt init_uart_clock**

3/ I have a CH304C on my Wemos 
**EDIT: i have a nodeMCU with CP2102 too**

Thanks !

**Comment by awawa-dev** [comment]
1) I haven't found one...the sk6812 RGBW protocol is very similar to WS281x it's technically possible but there are al least two obstacles:
- Arduino struggles to handle larger amount of LEDs and high serial baud-rate at the same time (500kb is already high speed). And for RGBW we need to send around ~33% more data to the LED strip.
- adalight is RGB only protocol, we need to provide RGB to RGBW transformations on the controller (Arduino/Esp). That sketch would be more complicated and drain more resources from the poor Arduino.
2) For HyperSerialEsp8266/HyperSerialEsp32/HyperSerialWled the speed is already hardcoded for 2Mb (and that speed you set up in HyperHDR) and the LED number in the strip is dynamic: it adjusts to the LED number sent from HyperHDR. You don't need to modify the sources and compile them for your ESP. 

**Please do not modify your init_uart_clock config.txt :wink: ! it's internal uart, you have independently handled by Kernel external USB serial adapters ch340g/cp2102...** 

3) There are special builds of HyperSerialWled working at @921600 speed but they are only for ESPs with serial chips not capable for 2mb. Your cp2102 can use it. CH340g should be capable of 2Mb if it's not a counterfeit (label on the chip should be visible).

**Reply by todely** [reply]
Ok i don't know why i wanna change UART speed of raspberry...stupid... i just have activate UART in config and speed is determined with hyperion config and sketch for ESP...

Thanks to you all seems ok RASP + ESP8266 + LED strip works **BUT** raspberry fail with error when activate usb capture with error in dmesg:

**EDIT: Raspberry PI 3**

[  115.519211] V4L2Worker invoked oom-killer: gfp_mask=0x400dc0(GFP_KERNEL_ACCOUNT|__GFP_ZERO), order=0, oom_score_adj=0
[  115.519230] CPU: 0 PID: 4456 Comm: V4L2Worker Tainted: G         C        5.4.51-v7+ #1333
[  115.519234] Hardware name: BCM2835
[  115.519238] Backtrace:
[  115.519259] [<8010d480>] (dump_backtrace) from [<8010d778>] (show_stack+0x20/0x24)
[  115.519268]  r6:b2820000 r5:00000000 r4:80d95cb8 r3:fd104aab
[  115.519281] [<8010d758>] (show_stack) from [<808cdbe4>] (dump_stack+0xe0/0x124)
[  115.519294] [<808cdb04>] (dump_stack) from [<8027b250>] (dump_header+0x64/0x200)
[  115.519304]  r8:80d05684 r7:b32eb9c0 r6:0000029e r5:b6776ac0 r4:b2821e00 r3:fd104aab
[  115.519315] [<8027b1ec>] (dump_header) from [<8027b8cc>] (oom_kill_process+0x140/0x188)
[  115.519326]  r9:80d97988 r8:80d05684 r7:80ab2954 r6:b2821e00 r5:b6777010 r4:b6776ac0
[  115.519337] [<8027b78c>] (oom_kill_process) from [<8027c448>] (out_of_memory+0x120/0x340)
[  115.519344]  r7:80d081c0 r6:80d04f48 r5:b2821e00 r4:80d081c0
[  115.519358] [<8027c328>] (out_of_memory) from [<802c85a8>] (__alloc_pages_nodemask+0x1144/0x123c)
[  115.519366]  r7:80d05048 r6:00000240 r5:00000000 r4:00000000
[  115.519379] [<802c7464>] (__alloc_pages_nodemask) from [<802adc18>] (__pte_alloc+0x34/0x1a0)
[  115.519389]  r10:00000800 r9:b4ce6c40 r8:00000040 r7:00000055 r6:80d04f48 r5:b2418b80
[  115.519393]  r4:b4ce6c00
[  115.519404] [<802adbe4>] (__pte_alloc) from [<802b1e00>] (handle_mm_fault+0xc80/0xd18)
[  115.519414]  r8:00000040 r7:00000055 r6:80d04f48 r5:b279cf00 r4:b279cf00 r3:00000000
[  115.519426] [<802b1180>] (handle_mm_fault) from [<808ef628>] (do_page_fault+0x320/0x3a8)
[  115.519437]  r10:00000800 r9:b4ce6c40 r8:00000805 r7:b4ce6c00 r6:2e174000 r5:b26ccc40
[  115.519441]  r4:b2821fb0
[  115.519453] [<808ef308>] (do_page_fault) from [<808ef76c>] (do_translation_fault+0xbc/0xc0)
[  115.519463]  r10:00000000 r9:75a3f258 r8:b2821fb0 r7:2e174000 r6:808ef6b0 r5:00000805
[  115.519467]  r4:00000805
[  115.519478] [<808ef6b0>] (do_translation_fault) from [<80115894>] (do_DataAbort+0x48/0xc4)
[  115.519486]  r6:808ef6b0 r5:00000805 r4:80d0a834 r3:00000005
[  115.519496] [<8011584c>] (do_DataAbort) from [<80101d24>] (__dabt_usr+0x44/0x60)
[  115.519501] Exception stack(0xb2821fb0 to 0xb2821ff8)
[  115.519509] 1fa0:                                     2e174000 000a9002 00000000 2e174008
[  115.519519] 1fc0: 6c900010 000a9000 0001517f 000a8c08 75a3f000 75a3f258 00000000 fffff000
[  115.519527] 1fe0: 75a3f26c 2d6cdaf0 75a3f258 759658d8 40000010 ffffffff
[  115.519536]  r8:10c5387d r7:10c5383d r6:ffffffff r5:40000010 r4:759658d8
[  115.519641] Mem-Info:
[  115.519661] active_anon:111084 inactive_anon:111155 isolated_anon:0
                active_file:111 inactive_file:249 isolated_file:0
                unevictable:4 dirty:0 writeback:0 unstable:0
                slab_reclaimable:2048 slab_unreclaimable:3123
                mapped:274 shmem:216 pagetables:787 bounce:0
                free:5393 free_pcp:391 free_cma:1765
[  115.519677] Node 0 active_anon:444336kB inactive_anon:444620kB active_file:444kB inactive_file:996kB unevictable:16kB isolated(anon):0kB isolated(file):0kB mapped:1096kB dirty:0kB writeback:0kB shmem:864kB writeback_tmp:0kB unstable:0kB all_unreclaimable? no
[  115.519698] Normal free:21572kB min:16384kB low:20480kB high:24576kB active_anon:444336kB inactive_anon:444620kB active_file:444kB inactive_file:1192kB unevictable:16kB writepending:0kB present:970752kB managed:948072kB mlocked:16kB kernel_stack:1144kB pagetables:3148kB bounce:0kB free_pcp:1564kB local_pcp:0kB free_cma:7060kB
[  115.519720] lowmem_reserve[]: 0 0
[  115.519735] Normal: 548*4kB (UMEC) 550*8kB (UMEC) 207*16kB (UMEC) 91*32kB (UMEC) 47*64kB (UEC) 15*128kB (UEC) 9*256kB (UEC) 1*512kB (E) 1*1024kB (E) 0*2048kB 0*4096kB = 21584kB
[  115.519820] 617 total pagecache pages
[  115.519831] 25 pages in swap cache
[  115.519851] Swap cache stats: add 30903, delete 30878, find 2571/3199
[  115.519858] Free swap  = 0kB
[  115.519865] Total swap = 102396kB
[  115.519872] 242688 pages RAM
[  115.519879] 0 pages HighMem/MovableOnly
[  115.519885] 5670 pages reserved
[  115.519892] 16384 pages cma reserved
[  115.519899] Tasks state (memory values in pages):
[  115.519907] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[  115.519942] [    108]     0   108     4742       48    61440      157             0 systemd-journal
[  115.519957] [    145]     0   145     4519        2    20480      212         -1000 systemd-udevd
[  115.519973] [    315]   100   315     5603        0    24576      166             0 systemd-timesyn
[  115.519985] [    356]   108   356     1483       47    14336       43             0 avahi-daemon
[  115.519997] [    358]     0   358     1987       14    16384       32             0 cron
[  115.520009] [    368] 65534   368     1080        5    14336       39             0 thd
[  115.520021] [    372]     0   372     6378        0    22528      162             0 rsyslogd
[  115.520033] [    377]   104   377     1669        1    16384      128          -900 dbus-daemon
[  115.520045] [    381]     0   381     6914        0    18432       28             0 rngd
[  115.520057] [    392]     0   392     3260        2    22528      160             0 systemd-logind
[  115.520070] [    393]     0   393     2685       12    20480       90             0 wpa_supplicant
[  115.520082] [    418]   108   418     1443        0    12288       63             0 avahi-daemon
[  115.520094] [    453]     0   453     2751       21    20480      125             0 wpa_supplicant
[  115.520107] [    480]     0   480      923        0    12288       52             0 alsactl
[  115.520120] [    489]     0   489      534        0    10240       28             0 hciattach
[  115.520132] [    492]     0   492     2384        0    16384       93             0 bluetoothd
[  115.520144] [    519]     0   519      726        0    12288      102             0 dhcpcd
[  115.520156] [    526]  1000   526   303003   221504  1089536    21332             0 hyperhdr
[  115.520168] [    533]     0   533     2681        0    18432      144         -1000 sshd
[  115.520180] [    538]     0   538     1654        0    16384       26             0 agetty
[  115.520192] [    539]     0   539     1077        0    12288       26             0 agetty
[  115.520204] [   1691]     0  1691     3059        2    20480      195             0 sshd
[  115.520217] [   2028]  1000  2028     3649        0    22528      230             0 systemd
[  115.520229] [   2034]  1000  2034     8817        0    30720      420             0 (sd-pam)
[  115.520241] [   2078]  1000  2078     3059       38    20480      158             0 sshd
[  115.520253] [   2087]  1000  2087     2123      143    16384       83             0 bash
[  115.520262] oom-kill:constraint=CONSTRAINT_NONE,nodemask=(null),cpuset=/,mems_allowed=0,global_oom,task_memcg=/,task=hyperhdr,pid=526,uid=1000
[  115.520368] Out of memory: Killed process 526 (hyperhdr) total-vm:1212012kB, anon-rss:886016kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:1064kB oom_score_adj:0
[  115.804754] oom_reaper: reaped process 526 (hyperhdr), now anon-rss:0kB, file-rss:912kB, shmem-rss:0kB

**Reply by awawa-dev** [reply]
Out of memory, there are few ways that could result in such scenario especially with 1GB but full logs (from the start to the crash...seems it's about 2 minutes, run hyperhdr with````/usr/bin/hyperhdr -d````) and backup of you configuration from the www general tab are needed. Try to lower capturing resolution & FPS. And use grabber option: 'quarter of frame' mode first to see it it helps.

**Reply by todely** [reply]
USB grabber was set to 640x480@30fps
With same configuration, quarter of frame seems to solve problem ! thanks !!!

**Reply by awawa-dev** [reply]
Quite strange, usually capturing using 1080p causes such problems.  Anyway it's good that works OK now.

**Reply by todely** [reply]
strange yes,i will try to start from scratch with raspberry

**Comment by todely** [comment]
Seems i need to add level shifter...to many V lost in cable from Wemos output to led strip...
could you please validate my schema :
![image](https://user-images.githubusercontent.com/11060244/138102663-cdcb98d7-a7f9-4afd-a83d-2b08dc808961.png)

**Comment by awawa-dev** [comment]
Diagram looks OK, but I didn't test such voltage level shifter, prefer SN74AHCT125N.
Are you sure it's a voltage problem because very often the lack of reaction from the LED strip lies in a specific ESP board not capable of 2Mb speed? Which firmware did you try and did you test WLED (to start some effects/color from its www panel directly)?

**Reply by todely** [reply]
Impossible to find 74AHCTx mounted on a board... does you know where to find one ? the only one i found seems to not be the best for led driver..
Voltage drop is clearly the problem, if i shorter cable it's ok...

**EDIT : Does this trick can work ?**
https://hackaday.com/2017/01/20/cheating-at-5v-ws2812-control-to-use-a-3-3v-data-line/

**Reply by awawa-dev** [reply]
Yes, it works also for sk6812 ;)
https://hyperhdr.blogspot.com/2020/12/my-build-log-using-sk6812-rgbw-led.html#chapter3

**Reply by todely** [reply]
Does the LED have to be the first one of the led strip (with first adress) or i can use any LED of led strip ?

**Comment by awawa-dev** [comment]
> Does the LED have to be the first one of the led strip (with first adress) or i can use any LED of led strip ?

You can use any of the LED from the strip: they are identical, identification is based only after the pass-through.

> Impossible to find 74AHCTx mounted on a board... does you know where to find one ? the only one i found seems to not be the best for led driver..

Alternatively for SN74AHCT125N need only universal board and dip socket (both under $1 total). And maybe some pins but you should receive them with the ESP board.
My newest custom implementation SN74AHCT125N board for ESP32 dedicated for HyperSPI: exposed input SPI interface on one row and the output to voltage level shifter and ground on the second row. 
![voltage_level_shifter_SN74AHCT125N](https://user-images.githubusercontent.com/69086569/138136278-860eb6b5-8c5c-4238-b6ef-1d02b3979b13.jpg)

**Reply by todely** [reply]
the one with LED more easy for me ^^ and can be done without universal board
Does this is correct ?
![image](https://user-images.githubusercontent.com/11060244/138140965-84b2295a-0c28-446e-8d1f-a7f24e593cc2.png)

**Reply by awawa-dev** [reply]
It's rather non-standard channel order on the LED (usually is 5v,data,ground or reversed) but yes, it's correct.
If you use that type of voltage level shifter you must include first LED from the level shifter in the LED strip design in HyperHDR.

**Reply by todely** [reply]
Ok so in ESP sketch i set true for "skip first led" and in hyperHDR i add one led
I have order few diode 1N4007, thanks again for your help !
